ships/0037-shipwright-build-strategy-controller.md

```diff
@@ -0,0 +1,190 @@
+So that I can provide Shipwright Build Strategies to my developers
+
+
+#### Story 2
+
+As a Kubernetes cluster administrator
+I want to be able to install , with an url I provide, my own catalog of Shipwright Builds Strategies via the Shipwright Operator
+So that I can provide a custom subset of Shipwright Build Strategies to my developers
+
+### Implementation Notes
+
+#### API
+
+We propose to add a new custom resource  the `ShipwrightBuildStrategy`.
+
+```yaml
+apiVersion: operator.shipwright.io/v1alpha1
+kind: ShipwrightBuildStrategy
+metadata:
+  name: buildStrategy
+spec:
+  targetNamespace:
+  catalog:
+    url:
+```
+
+#### Controller
+
+Actually the Shipwright Operator uses controller-runtime package to bootstrap a manager 
+and setup the Shipwright Build controller within that manager.
+In the target implementation we will continue using controller-runtime
+We will split the controllers each acheiving a specific function in its own package:
+- shipwrightbuild controller to reconcile ShipwrightBuild CRD and installs Shipwright Build manifests
+- shipwrightbuilstrategy controller to reconcile  ShipwrightBuildStrategy CRD and installs Shipwright Build Strategy catalog.
+The main package will bootstrap a controller manager and adds all the controllers to the manager.
+We propose this layer for package organisation
+
+```go
+main.go 
+controllers
+  controller.go
+  add_shipwrightbuild.go
+  add_shipwrightbuildstrategy.go
+  shipwrightbuild
+    shipwrightbuild_controller.go
+  shipwrightbuildstragegy
+     shipwrightbuildstrategy_controller.go
+```
+
+the controller package handles the list of AddToManagerFuncs
+and add to the manager any controller that register itself at package init time
+
+```go
+// AddToManagerFuncs is a list of functions to add all Controllers to the Manager
+var AddToManagerFuncs []func(manager.Manager) error
+
+// AddToManager adds all Controllers to the Manager
+func AddToManager(m manager.Manager) error {
+	for _, f := range AddToManagerFuncs {
+		if err := f(m); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+```
+
+add_shipwrightbuild.go contains
+
+```go
+func init() {
+	// AddToManagerFuncs is a list of functions to create controllers and add them to a manager.
+	AddToManagerFuncs = append(AddToManagerFuncs, shipwrightbuild.Add)
+}
+
+```
+
+add_shipwrightbuildstrategy.go contains
+
+```go
+func init() {
+	// AddToManagerFuncs is a list of functions to create controllers and add them to a manager.
+	AddToManagerFuncs = append(AddToManagerFuncs, shipwrightbuildstrategy.Add)
+}
+
+```
+
+### Test Plan
+
+The implementation has to be tested on a `unit`, `integration` and `e2e` level to ensure correctness.
+
+### Release Criteria
+
+tbd
+
+### Risks and Mitigations
+
+The default strategy catalog or any url provided catalog can be unreachable if the kubernetes cluster needs
+- a proxy to access to the catalog url
+- it is a disconnected cluster and doesnt have any access to internet
```

Comment:
Long term, do we need to think about providing OCI artifact resolvers, like Tekton does with task catalogs? Mirroring images is a well established practice for network-constrained environments.